File: doc/api-examples/python/monitor-agent-connection.py
#
# Copyright Contributors to the Eclipse BlueChi project
#
# SPDX-License-Identifier: MIT-0

from typing import Dict

import dasbus.connection
from dasbus.loop import EventLoop
from dasbus.typing import Variant

loop = EventLoop()
bus = dasbus.connection.SystemMessageBus()

agent_props_proxy = bus.get_proxy(
    "org.eclipse.bluechi.Agent",
    "/org/eclipse/bluechi",
    "org.freedesktop.DBus.Properties",
)


def on_connection_status_changed(
    interface: str,
    changed_props: Dict[str, Variant],
    invalidated_props: Dict[str, Variant],
) -> None:
    con_status = changed_props["Status"].get_string()
    print(f"Agent status: {con_status}")


agent_props_proxy.PropertiesChanged.connect(on_connection_status_changed)

loop.run()

File: doc/api-examples/rust/monitor-agent-connection.rs
// Copyright Contributors to the Eclipse BlueChi project
//
// SPDX-License-Identifier: MIT-0

use dbus::{
    arg::Variant,
    blocking::{stdintf::org_freedesktop_dbus::PropertiesPropertiesChanged, Connection},
    Message,
};
use std::time::Duration;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let conn = Connection::new_system()?;

    let bluechi = conn.with_proxy(
        "org.eclipse.bluechi.Agent",
        "/org/eclipse/bluechi",
        Duration::from_millis(5000),
    );
    let _id = bluechi.match_signal(
        |signal: PropertiesPropertiesChanged, _: &Connection, _: &Message| {
            match signal.changed_properties.get_key_value("Status") {
                Some((_, Variant(changed_value))) => {
                    println!("Agent status: {}", changed_value.as_str().unwrap_or(""))
                }
                _ => {}
            }
            true
        },
    );

    loop {
        conn.process(Duration::from_millis(1000))?;
    }
}
